### PR TITLE
RecordsWriter, Publisher, and Search 

### PR DESCRIPTION
**Pull request by DallanQ**
Lots going on here: I had three branches that I was going to issue three separate pull requests for, but they build upon one another and in the end I decided to just submit one PR for the final branch. This PR has everything needed to upload a CSV file, publish it (index it), and search the published records. 

It requires deleting the cms_esdata volume, creating a new one, and running es_setup.sh from the elasticsearch directory (instructions in the readme).

New processes:
1. recordswriter
2. publisher
3. search server

Questions:
1. Should the search server be combined with server? 
2. If so, is it worth it to create a separate lambda executable? Search will be called probably 100x more often than the other CMS functions, but there isn't that much extra code in server, and I'm wondering if my making search a separate server is a case of premature optimization.

ToDos:
1. put swagger notes on the search server functions

### CODE REVIEW

**Comment by jancona** [comment]
> Questions:
> 
> 1. Should the search server be combined with server?
Having them separate adds some deploy complexity. 
> 2. If so, is it worth it to create a separate lambda executable? Search will be called probably 100x more often than the other CMS functions, but there isn't that much extra code in server, and I'm wondering if my making search a separate server is a case of premature optimization.

Just to make sure we're on the same page, I'm going to be a bit pedantic. Right now you have two server processes, the API server on port 3000 and the search server on 3001. These process are separate Go executables. So a client goes to http://localhost:3000/collections or to http://localhost:3001/search. You could combine them into a single executable and run two instances, one on port 3000 and the other on 3001 and clients couldn't tell the difference. But if you want clients to be able to go to http://localhost:3000/collections and http://localhost:3000/search, the endpoints either must be in a single executable or you have to run a reverse proxy like NGINX in front of the separate executables.

API Gateway functions sort of like NGINX, so you can run separate executables while keeping a single base URL. There's no reason I can think of to have a separate API Gateway API for search, since it allows [throttling by route](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-protect.html). 

It looks like the search server still needs a Postgres connection. If it didn't, that would be a reason to keep it in a separate Lambda (although you could still have a single Go executable if you wanted). So I don't see a strong argument either way right now.

**Comment by DallanQ** [comment]
Ok, I will combine search into the main server Monday morning. Originally I
was thinking that the search server wouldn’t use much of the existing code,
but when I put the search functionality into the API object and called the
database to read the records, which I think is a good idea, that turned out
not to be the case. I don’t expect that the same client would call both
sets of endpoints, but splitting the servers just introduces complexity.

Thanks for the feedback.

Dallan

**Comment by jancona** [comment]
Sounds good. I've started a review on the rest of the code. I'll try to finish it later today.

File: api/collections.go [hunk 1]
```diff
@@ -25,6 +25,15 @@ func (api API) GetCollections(ctx context.Context /* filter/search criteria */)
 	return &CollectionResult{Collections: cols}, nil
 }
 
+// GetManyCollections holds the business logic around getting many Collections
+func (api API) GetManyCollections(ctx context.Context, ids []string) ([]model.Collection, *model.Errors) {
+	colls, err := api.collectionPersister.SelectManyCollections(ctx, ids)
```
[the remark this answers is not in the document]
Comment (DallanQ) [reply api/collections.go:30]: I like GetCollectionsByID (and GetRecordsByID) - I'll change it

File: publisher/server.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	ctx := context.Background()
+
+	// parse environment
+	env, err := ParseEnv()
+	if err != nil {
+		log.Fatalf("[FATAL] Error parsing environment variables: %v", err)
+	}
+
+	filter := &logutils.LevelFilter{
+		Levels:   []logutils.LogLevel{"DEBUG", "INFO", "ERROR", "FATAL"},
```
[the remark this answers is not in the document]
Comment (DallanQ) [reply publisher/server.go:67]: That makes sense. I've converted the warnings to either errors or infos. Thanks for the pointer.

File: api/api.go [hunk 3]
```diff
@@ -100,53 +152,92 @@ func (api *API) PostPersister(cp model.PostPersister) *API {
 	return api
 }
 
+// PostPersister sets the PostPersister for the api
```
[the remark this answers is not in the document]
Comment (DallanQ) [reply api/api.go:155]: thanks for the catch.

File: api/search.go [hunk 4]
```diff
@@ -0,0 +1,1107 @@
...
+		} else {
+			// Print the response status and error information.
+			msg := fmt.Sprintf("[%s] %#v\n", res.Status(), e)
+			log.Println(msg)
+			return nil, model.NewErrors(http.StatusInternalServerError, errors.New(msg))
+		}
+	}
+
+	// get hit data
+	var r map[string]interface{}
```
[the remark this answers is not in the document]
Comment (DallanQ) [reply api/search.go:333]: That's a good point. I started out that way for convenience, but now that I know what I need to extract from the responses, I should decode the responses into structs. I'll do that.

**Pull request merged**